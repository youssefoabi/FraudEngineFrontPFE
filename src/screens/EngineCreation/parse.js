import { isNil, map, pathOr, pipe, values } from 'ramda';
import { isNilOrEmpty } from 'ramda-adjunct';

import { FIELDS_BY_ID } from '../../components/SearchCriteriaSettings/components/FieldsBlock/config';

const parse = engine => {
  if (isNil(engine)) return null;

  const {
    id,
    name,
    isEnable,
    scopes,
    inputFields,
    backgroundImages,
    marketingText,
    isMarketingTextEnable,
    searchText,
    engineId,
    createdAt,
    modifiedAt,
  } = engine;

  const logo = {
    urlImageDesktop: 'string',
    urlLinkDesktop: 'string',
    urlImageMobile: 'string',
    urlLinkMobile: 'string',
    order: 1,
    isEnable: true,
    alt: 'string',
    openInNewTab: true,
  };

  const mapParameter = param => {
    if (isNilOrEmpty(param)) return null;

    const { scopeId: scopeParameterId, key: label } = param || {};
    return { ...param, scopeParameterId, label };
  };

  const mapInputFields = allInputFields => {
    if (isNilOrEmpty(allInputFields)) return null;

    return pipe(
      values,
      map(inputField => {
        const type = pathOr(null, ['type', 'value'], inputField);
        const parameters = pathOr([], ['parameters'], inputField);
        const { label } = inputField || {};
        const newParameters = map(mapParameter, parameters);
        return {
          ...(inputField || {}),
          type,
          fieldTypeId: FIELDS_BY_ID[type],
          inputFieldLabel: label,
          parameters: newParameters,
        };
      }),
    )(allInputFields);
  };

  return {
    engineCode: id,
    code: id,
    name,
    isEnable,
    scopes: isNilOrEmpty(scopes) ? [] : values(scopes),
    inputFields: isNilOrEmpty(inputFields) ? [] : mapInputFields(inputFields),
    backgroundImages: isNilOrEmpty(backgroundImages) ? [] : values(backgroundImages),
    marketingText: {
      text: marketingText,
      isEnable: isMarketingTextEnable,
    },
    searchText,
    engineId,
    logo,
    createdAt,
    modifiedAt,
  };
};
export default parse;
