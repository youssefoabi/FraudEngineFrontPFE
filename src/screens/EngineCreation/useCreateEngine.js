import { useCallback } from 'react';
import { useDispatch } from 'react-redux';

import { CREATE_ENGINE_REQUEST } from '../../components/EngineForm/actions';

export default function useCreateEngine() {
  const dispatch = useDispatch();

  const createEngine = useCallback(engine => dispatch({ type: CREATE_ENGINE_REQUEST, payload: { engine } }), [
    dispatch,
  ]);

  return { createEngine };
}
