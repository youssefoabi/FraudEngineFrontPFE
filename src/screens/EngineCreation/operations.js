import { call, put, takeLatest } from 'redux-saga/effects';
import { isNilOrEmpty } from 'ramda-adjunct';

import {
  CREATE_ENGINE_ERROR,
  CREATE_ENGINE_REQUEST,
  CREATE_ENGINE_SUCCESS,
} from '../../components/EngineForm/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';
import parse from './parse';

export function* createEngine({ payload }) {
  yield put({ type: PENDING_REQUEST });

  try {
    const { engine } = payload;
    if (isNilOrEmpty(engine)) return null;

    yield call(clientApi().createEngine, parse(engine));

    yield put({ type: RESET_PENDING_REQUEST });
    yield put({ type: CREATE_ENGINE_SUCCESS });
  } catch (error) {
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({
      type: CREATE_ENGINE_ERROR,
      error: { errorMessage: "Erreur d'enregistrement du moteur de recherche" },
    });
  }

  return true;
}

export default [takeLatest(CREATE_ENGINE_REQUEST, createEngine)];
