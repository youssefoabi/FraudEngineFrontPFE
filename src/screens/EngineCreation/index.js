import React, { useState } from 'react';

import EngineScreen from '../EngineScreen';
import useCreateEngine from './useCreateEngine';

const initialState = {
  isEnable: false,
  inputFields: {
    0: {
      id: 0,
      isEnable: false,
      isMandatory: false,
      parameters: [],
    },
    1: {
      id: 1,
      isEnable: false,
      isMandatory: false,
      parameters: [],
    },
  },
  backgroundImages: {
    0: { isEnable: false, openInNewTab: false },
  },
  isMarketingTextEnable: false,
};

function EngineCreation() {
  const [engine, setEngine] = useState(initialState);

  const { createEngine } = useCreateEngine();

  return <EngineScreen engine={engine} action={createEngine} title="CrÃ©ation" setEngine={setEngine} />;
}

export default EngineCreation;
