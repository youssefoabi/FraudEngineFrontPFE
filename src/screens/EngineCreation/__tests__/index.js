import { itRendersCorrectly } from '@cdiscount/jest-utils';
import React from 'react';

import EngineCreation from '../';

jest.mock('react-beautiful-dnd', () => ({
  Droppable: ({ children }) =>
    children(
      {
        draggableProps: {
          style: {},
        },
        innerRef: jest.fn(),
      },
      {},
    ),
  Draggable: ({ children }) =>
    children(
      {
        draggableProps: {
          style: {},
        },
        innerRef: jest.fn(),
      },
      {},
    ),
  DragDropContext: ({ children }) => children,
}));

jest.mock('react-redux', () => ({
  useDispatch: () => ({}),
}));

describe(EngineCreation.name, () => {
  itRendersCorrectly(<EngineCreation />);
});
