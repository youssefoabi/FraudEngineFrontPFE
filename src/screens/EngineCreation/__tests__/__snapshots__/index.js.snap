// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EngineCreation renders correctly 1`] = `
<EngineScreen
  action={[Function]}
  engine={
    Object {
      "backgroundImages": Object {
        "0": Object {
          "isEnable": false,
          "openInNewTab": false,
        },
      },
      "inputFields": Object {
        "0": Object {
          "id": 0,
          "isEnable": false,
          "isMandatory": false,
          "parameters": Array [],
        },
        "1": Object {
          "id": 1,
          "isEnable": false,
          "isMandatory": false,
          "parameters": Array [],
        },
      },
      "isEnable": false,
      "isMarketingTextEnable": false,
    }
  }
  setEngine={[Function]}
  title="CrÃ©ation"
/>
`;
